plugin/cache/handler.go

```diff
@@ -25,37 +26,35 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 
 	do := state.Do() // TODO(): might need more from OPT record? Like the actual bufsize?
 
-	now := time.Now().UTC()
+	now := c.now().UTC()
 
 	i, ttl := c.get(now, qname, qtype, do)
 	if i != nil && ttl > 0 {
-		resp := i.toMsg(r)
+		resp := i.toMsg(r, now)
 
 		state.SizeAndDo(resp)
 		resp, _ = state.Scrub(resp)
 		w.WriteMsg(resp)
 
 		if c.prefetch > 0 {
 			i.Freq.Update(c.duration, now)
-		}
-
-		pct := 100
-		if i.origTTL != 0 { // you'll never know
-			pct = int(float64(ttl) / float64(i.origTTL) * 100)
-		}
-
-		if c.prefetch > 0 && i.Freq.Hits() > c.prefetch && pct < c.percentage {
-			// When prefetching we loose the item i, and with it the frequency
-			// that we've gathered sofar. See we copy the frequencies info back
-			// into the new item that was stored in the cache.
-			prr := &ResponseWriter{ResponseWriter: w, Cache: c, prefetch: true}
-			plugin.NextOrFailure(c.Name(), c.Next, ctx, prr, r)
 
-			if i1, _ := c.get(now, qname, qtype, do); i1 != nil {
-				i1.Freq.Reset(now, i.Freq.Hits())
+			threshold := int(math.Ceil(float64(c.percentage) / 100 * float64(i.origTTL)))
+			if i.Freq.Hits() >= c.prefetch && ttl <= threshold {
+				go func() {
```

Comment:
(previous comment is gone?)

still don't see why this is needing to run in a goroutine?

Reply:
Is the commit message not helpful?

> The logging and metrics plugins (among others) included the duration of
a cache prefetch in the request latency of client request. This change
fixes this wrong reporting and executes the prefetch request in a
goroutine in the background.

The prefetch should not be counted towards the client request latency, as that can trip alerts on high latency.

Reply:
This doesn't matter. Other plugins like log and metrics wait for the termination of `ServeDNS()` (called by `NextOrFailure`) to determine the request latency.